### PR TITLE
Publish artifacts to Maven Central using Gradle

### PR DESCRIPTION
**Pull request by tabboud**
## Before this PR
Artifacts were only published to Github releases and there was no way for Java projects to consume the artifacts.

## After this PR

==COMMIT_MSG==
Publish artifacts to Maven Central using Gradle
==COMMIT_MSG==

This project will now publish all artifacts (`go-java-launcher` and `go-init`) to Maven Central using Gradle and the gradle-external-publish plugin. This is achieved through the OSS circle template variable which configures circle to restore the cached artifacts, built via Godel in the `dist` step, and uses Gradle to publish them, as-is, to Sonatype.

### CODE REVIEW

**Comment by tabboud** [comment]
This has not yet published successfully from my branch, which I believe is because this is not a tagged build.
```
> Task :publishDistPublicationToSonatypeRepository SKIPPED
Skipping task ':publishDistPublicationToSonatypeRepository' as task onlyIf is false.
```

File: .circleci/template.sh [hunk 1]
```diff
@@ -1,2 +1,3 @@
 #!/usr/bin/env bash
 export CIRCLECI_TEMPLATE=go-distribution-oss
+export PUBLISH_WITH_GRADLE=true
```
Comment (tabboud) [inline .circleci/template.sh:3]: See the internal https://github.p.b/excavator-checks/circle-templates/pull/900/ for where this is added.

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+apply plugin: 'com.palantir.external-publish-custom'
+apply plugin: 'com.palantir.git-version'
+
+group 'com.palantir.launching'
+
+version ((System.env.CIRCLE_TAG ?: gitVersion()).replaceAll('^v', ''))
+
+externalPublishing {
+    publication('dist') {
+        artifact file('out/dist/go-java-launcher/${project.version}/bin/go-java-launcher-${project.version}.tgz')
```
Comment (CRogers) [inline build.gradle]: These need `"` quotes so it will interpolate the `${project.version}`.

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        artifact file('out/dist/go-init/${project.version}/bin/go-init-${project.version}.tgz')
```
Comment (CRogers) [inline build.gradle]: Hmm, to publish two things I think we're going to need two publications, with different names

Comment (CRogers) [reply build.gradle]:
```
# CIRCLE_TAG=1.12.0-10-g0ac11f6 ./gradlew publishToMavenLocal
> Task :publishDistPublicationToMavenLocal FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':publishDistPublicationToMavenLocal'.
> Failed to publish publication 'dist' to repository 'mavenLocal'
   > Invalid publication 'dist': multiple artifacts with the identical extension and classifier ('tgz.asc', 'null').
```

**Comment by CRogers** [comment]
Ok, I pushed up some changes and I think this should work now I've tested it with maven local:

```console
# tree ~/.m2/repository/com/palantir/
/root/.m2/repository/com/palantir/
`-- launching
    |-- go-init
    |   |-- 1.12.0-10-g0ac11f6
    |   |   |-- go-init-1.12.0-10-g0ac11f6.pom
    |   |   |-- go-init-1.12.0-10-g0ac11f6.pom.asc
    |   |   |-- go-init-1.12.0-10-g0ac11f6.tgz
    |   |   `-- go-init-1.12.0-10-g0ac11f6.tgz.asc
    |   `-- maven-metadata-local.xml
    `-- go-java-launcher
        |-- 1.12.0-10-g0ac11f6
        |   |-- go-java-launcher-1.12.0-10-g0ac11f6.pom
        |   |-- go-java-launcher-1.12.0-10-g0ac11f6.pom.asc
        |   |-- go-java-launcher-1.12.0-10-g0ac11f6.tgz
        |   `-- go-java-launcher-1.12.0-10-g0ac11f6.tgz.asc
        `-- maven-metadata-local.xml
```

**Comment by CRogers** [comment]
:+1:

**Review by jkozlowski — APPROVED** [review]

**Pull request merged**